engine/table/src/main/java/io/deephaven/engine/table/impl/locations/impl/AbstractTableLocationProvider.java

```diff
@@ -232,14 +328,12 @@ protected void doInitialization() {
     @Override
     @NotNull
     public final Collection<ImmutableTableLocationKey> getTableLocationKeys() {
-        // We need to prevent reading the map (and maybe mutating it?) during a transaction.
-        // We could transition between two maps, a stable copy and a shadow copy that is being mutated.
-        // Or we could hold a bigger lock while mutating the map, and hold the same lock here. Sounds like a job for a
-        // read-write lock (e.g. ReentrantReadWriteLock), maybe. If you want `FunctionalLock`, the pattern (but mostly
-        // not the code) from io.deephaven.engine.updategraph.UpdateGraphLock could help.
-        // I think we need the read-write lock for correctness, and I think we need to make it explicit. That is, the
-        // user needs to be able to get a read lock and hold it while it's operating on the returned collection.
-        return unmodifiableTableLocationKeys;
+        // This lock is held while `endTransaction()` updates `tableLocations` with the accumulated adds/removes.
+        // Locking here ensures that this call won't return while `tableLocations` (and `unmodifiableTableLocationKeys`)
+        // contain a partial transaction.
+        synchronized (tableLocations) {
+            return unmodifiableTableLocationKeys;
+        }
```

Comment:
You need to return a copy. Maybe audit usages to make sure we don't do this more than once per table?